File: pipeline/detectors/make_missing_target.py
"""
Detector for make missing target errors.
"""

import re
import typing as T
from pipeline.detectors.base import Detector
from pipeline.models import ErrorClue


class MakeMissingTargetDetector(Detector):
    """
    Detect make errors when a required source file is missing.

    Matches patterns like:
    - make: *** No rule to make target 'dim.c', needed by 'dim'.  Stop.
    - make[1]: *** No rule to make target 'src/file.c', needed by 'target'.  Stop.
    """


    PATTERNS = {
        "make_missing_target": r"No rule to make target '(?P<target>[^']+)', needed by '(?P<needed_by>[^']+)'",
    }

    EXAMPLES = [
        (
            "make: *** No rule to make target 'dim.c', needed by 'dim'.  Stop.",
            {
                "clue_type": "make_missing_target",
                "confidence": 1.0,
                "context": {
                    "target": "dim.c",
                    "needed_by": "dim",
                },
            },
        ),
        (
            "make[1]: *** No rule to make target 'src/file.c', needed by 'target'.  Stop.",
            {
                "clue_type": "make_missing_target",
                "confidence": 1.0,
                "context": {
                    "target": "src/file.c",
                    "needed_by": "target",
                },
            },
        ),
    ]

File: src/boil/detectors/make_missing_target.go
// Detector for make missing target errors.
package detectors

// MakeMissingTargetDetector detects make missing target errors
type MakeMissingTargetDetector struct {
	*BaseDetector
}

func NewMakeMissingTargetDetector() (*MakeMissingTargetDetector, error) {
	base, err := NewBaseDetector(
		"MakeMissingTargetDetector",
		100,
		map[string]string{
			"make_missing_target": `No rule to make target '(?P<target>[^']+)', needed by '(?P<needed_by>[^']+)'`,
		},
		[]DetectorExample{
			{
				Name:     "make_missing_target",
				Input:    "No rule to make target 'utils.o', needed by 'main'",
				ClueType: "make_missing_target",
				Context: map[string]string{
					"target":    "utils.o",
					"needed_by": "main",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &MakeMissingTargetDetector{base}, nil
}
